third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs

```diff
@@ -0,0 +1,348 @@
+}
+
+impl AvailCopiesAnalysis {
+    /// Create a new instance of definitely available copies analysis.
+    /// `code` is the bytecode of the function being analyzed.
+    pub fn new(code: &[Bytecode]) -> Self {
+        Self {
+            borrowed_locals: Self::get_borrowed_locals(code),
+        }
+    }
+
+    /// Analyze the given function and return the definitely available copies annotation.
+    fn analyze(&self, func_target: &FunctionTarget) -> AvailCopiesAnnotation {
+        let code = func_target.get_bytecode();
+        let cfg = StacklessControlFlowGraph::new_forward(code);
+        let block_state_map = self.analyze_function(AvailCopies::new(), code, &cfg);
+        let per_bytecode_state =
+            self.state_per_instruction(block_state_map, code, &cfg, |before, after| {
+                AvailCopiesState {
+                    before: before.clone(),
+                    after: after.clone(),
+                }
+            });
+        AvailCopiesAnnotation(per_bytecode_state)
+    }
+
+    /// Get the set of locals that have been borrowed in the function being analyzed.
+    fn get_borrowed_locals(code: &[Bytecode]) -> BTreeSet<TempIndex> {
+        code.iter()
+            .filter_map(|bc| {
+                if let Bytecode::Call(_, _, Operation::BorrowLoc, srcs, _) = bc {
+                    // BorrowLoc should have only one source.
+                    srcs.first().cloned()
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+}
+
+impl TransferFunctions for AvailCopiesAnalysis {
+    type State = AvailCopies;
+
+    // This is a forward analysis.
+    const BACKWARD: bool = false;
+
+    fn execute(&self, state: &mut Self::State, instr: &Bytecode, _offset: CodeOffset) {
+        use Bytecode::*;
+        instr.dests().iter().for_each(|dst| {
+            state.kill_copies_with(*dst, &self.borrowed_locals);
+        });
+        if let Assign(_, dst, src, _) = instr {
+            state.make_copy_available(*dst, *src, &self.borrowed_locals);
+        }
+    }
+}
+
+impl DataflowAnalysis for AvailCopiesAnalysis {}
+
+/// Processor for the definitely available copies analysis.
+pub struct AvailCopiesAnalysisProcessor();
+
+impl FunctionTargetProcessor for AvailCopiesAnalysisProcessor {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        func_env: &FunctionEnv,
+        mut data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if func_env.is_native() {
+            return data;
+        }
+        let target = FunctionTarget::new(func_env, &data);
+        let analysis = AvailCopiesAnalysis::new(target.get_bytecode());
+        let annotation = analysis.analyze(&target);
+        data.annotations.set(annotation, true);
+        data
+    }
+
+    fn name(&self) -> String {
+        "available_copies".to_string()
```

Comment:
Please check the other processors, but I think the names are currently caml-cased.

Reply:
I may have named this based on how analyses under move-model/bytecode are named (eg, livevar_analysis.rs), however, the ones under compiler v2 pipeline are named using caml case.

Changing these for consistency.